README.md

```diff
@@ -695,200 +775,441 @@ Tuples, lists, and strings are also monoids:
+
+# Example usage:
+a = Setoid(1)
+b = Setoid(1)
+c = Setoid(2)
+
+assert a.equals(b)
+assert not a.equals(c)
 ```
 
-## Semigroup (TODO)
+## Semigroup
 
-An object that has a `concat` function that combines it with another object of the same type.
+A semigroup is an algebraic structure with a `concat` method that combines it with another object of the same type.
 
 ```python
-# TODO
+class Semigroup:
+    def __init__(self, value):
+        self.value = value
+
+    def concat(self, other):
+        return Semigroup(self.value + other.value)
+
+# Example usage:
+a = Semigroup("Hello, ")
+b = Semigroup("World!")
+
+assert a.concat(b).value == "Hello, World!"
 ```
 
-## Foldable (TODO)
+## Foldable
 
-An object that has a `reduce` function that applies a function against an accumulator and each element in the array (from left to right) to reduce it to a single value.
+A foldable is an object that has a `reduce` method that applies a function against an accumulator and each element in the object to reduce it to a single value.
 
 ```python
-# TODO
+class Foldable(list):
+    def reduce(self, function, initial):
+        accumulator = initial
+        for value in self:
+            accumulator = function(accumulator, value)
+        return accumulator
+
+# Example usage:
+numbers = Foldable([1, 2, 3, 4])
+sum_of_numbers = numbers.reduce(lambda acc, x: acc + x, 0)
+
+assert sum_of_numbers == 10
 ```
 
-## Lens (TODO)
+## Lens
 
-A lens is a structure (often an object or function) that pairs a getter and a non-mutating setter for some other data
-structure.
+A lens is a composable structure that pairs a getter and a setter for an immutable update of a data structure.
 
 ```python
-# TODO
+class Lens:
+    def __init__(self, getter, setter):
+        self._getter = getter
+        self._setter = setter
+
+    def view(self, data):
+        return self._getter(data)
+
+    def set(self, data, value):
+        return self._setter(data, value)
+
+    def over(self, data, function):
+        return self.set(data, function(self.view(data)))
+
+# Example usage:
+name_lens = Lens(lambda data: data['name'], lambda data, value: {**data, 'name': value})
+person = {'name': 'Alice', 'age': 25}
+
+assert name_lens.view(person) == 'Alice'
+assert name_lens.set(person, 'Bob') == {'name': 'Bob', 'age': 25}
 ```
 
-Lenses are also composable. This allows easy immutable updates to deeply nested data.
+Lenses can be composed to focus on nested data structures.
 
 ```python
-# TODO
+# Composing lenses to focus on nested structures
+address_lens = Lens(lambda data: data['address'], lambda data, value: {**data, 'address': value})
+street_lens = Lens(lambda data: data['street'], lambda data, value: {**data, 'street': value})
+
+full_address_lens = address_lens.compose(street_lens)
+person_with_address = {'name': 'Alice', 'address': {'street': '123 Main St', 'city': 'Anytown'}}
+
+assert full_address_lens.view(person_with_address) == '123 Main St'
```

Comment:
Please, mention https://github.com/mahmoud/glom here, it is a nice reference implementation